akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala

```diff
@@ -662,28 +724,42 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
 
   override def shutdown(): Future[Done] = {
     _shutdown = true
-    killSwitch.shutdown()
-    topLevelFREvents.loFreq(Transport_KillSwitchPulled, NoMetaData)
-    if (taskRunner != null) {
-      taskRunner.stop()
-      topLevelFREvents.loFreq(Transport_Stopped, NoMetaData)
-    }
-    if (aeronErrorLogTask != null) {
-      aeronErrorLogTask.cancel()
-      topLevelFREvents.loFreq(Transport_AeronErrorLogTaskStopped, NoMetaData)
-    }
-    if (aeron != null) aeron.close()
-    if (mediaDriver.isDefined) {
-      stopMediaDriver()
-      topLevelFREvents.loFreq(Transport_MediaFileDeleted, NoMetaData)
+    val allAssociations = associationRegistry.allAssociations
+    val flushing: Future[Done] =
+      if (allAssociations.isEmpty) Future.successful(Done)
+      else {
+        val flushingPromise = Promise[Done]()
+        system.systemActorOf(FlushOnShutdown.props(flushingPromise, shutdownFlushTimeout,
+          this, allAssociations).withDispatcher(remoteSettings.Dispatcher), "remoteFlushOnShutdown")
+        flushingPromise.future
+      }
+    implicit val ec = remoteDispatcher
+    flushing.recover { case _ ⇒ Done }.map { _ ⇒
+      killSwitch.shutdown()
+
+      topLevelFREvents.loFreq(Transport_KillSwitchPulled, NoMetaData)
+      if (taskRunner != null) {
+        taskRunner.stop()
+        topLevelFREvents.loFreq(Transport_Stopped, NoMetaData)
+      }
+      if (aeronErrorLogTask != null) {
+        aeronErrorLogTask.cancel()
+        topLevelFREvents.loFreq(Transport_AeronErrorLogTaskStopped, NoMetaData)
+      }
+      if (aeron != null) aeron.close()
+      if (mediaDriver.isDefined) {
+        stopMediaDriver()
+        topLevelFREvents.loFreq(Transport_MediaFileDeleted, NoMetaData)
+      }
+      topLevelFREvents.loFreq(Transport_FlightRecorderClose, NoMetaData)
+
+      flightRecorder.foreach(_.close())
```

Comment:
@drewhk it seems like it's this `close` that causes the seg fault. Probably this PR changes the timing so that some events are still concurrently being written when this `close` is done.